src/utils/parse.rs

```diff
@@ -0,0 +1,186 @@
+		}
+    }
+}
+
+/// Look up a guild member by a string case-insensitively.
+///
+/// Requires the cache feature to be enabled.
+///
+/// The lookup strategy is as follows (in order):
+/// 1. Lookup by ID.
+/// 2. Lookup by mention.
+/// 3. Lookup by name#discrim
+/// 4. Lookup by name
+/// 5. Lookup by nickname
+#[cfg(feature = "cache")]
+#[async_trait::async_trait]
+impl Parse for Member {
+    type Err = MemberParseError;
+
+    #[doc = ""] // to get impl documentation to show
+    async fn parse(ctx: &Context, msg: &Message, s: &str) -> Result<Self, Self::Err> {
+        let guild = msg.guild(&ctx.cache).await.ok_or(MemberParseError::GuildNotInCache)?;
+
+        let lookup_by_id = || guild.members.get(&UserId(s.parse().ok()?));
+
+        let lookup_by_mention = || {
+            guild.members.get(&UserId(
+                s.strip_prefix("<@")?.trim_start_matches('!').strip_suffix(">")?.parse().ok()?,
+            ))
+        };
+
+        let lookup_by_name_and_discrim = || {
+            let pound_sign = s.find('#')?;
+            let name = &s[..pound_sign];
+            let discrim = s[(pound_sign + 1)..].parse::<u16>().ok()?;
+            guild.members.values().find(|member| {
+                member.user.discriminator == discrim && member.user.name.eq_ignore_ascii_case(name)
+            })
+        };
+
+        let lookup_by_name = || guild.members.values().find(|member| member.user.name == s);
+
+        let lookup_by_nickname = || {
+            guild.members.values().find(|member| match &member.nick {
+                Some(nick) => nick.eq_ignore_ascii_case(s),
+                None => false,
+            })
+        };
+
+        lookup_by_id()
+            .or_else(lookup_by_mention)
+            .or_else(lookup_by_name_and_discrim)
+            .or_else(lookup_by_name)
+            .or_else(lookup_by_nickname)
+            .cloned()
+            .ok_or(MemberParseError::NotFoundOrMalformed)
+    }
+}
+
+/// Error that can be returned from [`Message::parse`].
+#[derive(Debug)]
+pub enum MessageParseError {
+    Malformed,
+    Http(SerenityError),
+    /// When the `gateway` feature is disabled and the required information was not in cache
+    HttpNotAvailable,
+}
+
+impl std::error::Error for MessageParseError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::Malformed => None,
+            Self::Http(e) => Some(e),
+            Self::HttpNotAvailable => None,
+        }
+    }
+}
+
+impl std::fmt::Display for MessageParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Malformed => {
+                write!(f, "provided string did not adhere to any known guild message format")
+            },
+            Self::Http(e) => write!(f, "failed to request message data via HTTP: {}", e),
```

Comment:
I see, I wasn't sure which casing should be preferred for error messages inside serenity.

It may be of interest that even existing code isn't fully consistent about this: some code uses uppercase errors, some lowercase:
![Screenshot_20210223_201001](https://user-images.githubusercontent.com/21220820/108895009-50fac500-7613-11eb-8e60-dd5c8c6d60b3.png)

Reply:
Another thing: with the convention being to add a trailing full stop to error messages, wouldn't `{}.` expand to two full stops? Think `Failed to request message data via HTTP: Some HTTP specific error..` (note the `..`)